File: 0015-3sum/three-pointer.cpp
#include <algorithm>
#include <vector>

using namespace std;

class Solution
{
    public:
        vector<vector<int>> threeSum(vector<int>& nums)
        {
            // Initialize answer vector
            vector<vector<int>> result;

            // Sort the vector
            sort(nums.begin(), nums.end());

            for (int i = 0; i < nums.size() - 2; i++)
            {
                // Skip duplicates
                if ((i > 0) && (nums[i] == nums[i - 1]))
                {
                    continue;
                }
                else
                {
                    int j = i + 1;
                    int k = nums.size() - 1;

                    while (j < k)
                    {
                        int sum = nums[i] + nums[j] + nums[k];

                        if (sum == 0)
                        {
                            result.push_back({nums[i], nums[j], nums[k]});
                            k--;

                            // Skip duplicates
                            while ((j < k) && (nums[k] == nums[k + 1]))
                            {
                                k--;
                            }
                        }
                        else if (sum > 0)
                        {
                            k--;
                        }
                        else
                        {
                            j++;
                        }
                    }
                }
            }

            return result;    
        }
};

File: 0015-3sum/three-pointer.rs
impl Solution 
{
    pub fn three_sum(nums: Vec<i32>) -> Vec<Vec<i32>>
    {
        let n: usize = nums.len() as usize;

        // Sort array
        let mut arr: Vec<i32> = nums;
        arr.sort();

        // Initialize return vector
        let mut result: Vec<Vec<i32>> = Vec::new();

        for i in 0..(n-2)
        {
            // Skip duplicates
            if ((i > 0) && (arr[i] == arr[i - 1]))
            {
                continue;
            }
            else
            {
                let mut j: usize = (i + 1) as usize;
                let mut k: usize = (n - 1) as usize;

                while (j < k)
                {
                    let sum: i32 = arr[i] + arr[j] + arr[k];

                    if (sum == 0)
                    {
                        // Add triplet
                        let tmp: Vec<i32> = vec![arr[i], arr[j], arr[k]];
                        result.push(tmp);

                        k -= 1;

                        // Skip duplicates
                        while ((j < k) && (arr[k] == arr[k + 1]))
                        {
                            k -= 1;
                        }
                    }
                    else if (sum > 0)
                    {
                        k -= 1;
                    }
                    else
                    {
                        j += 1;
                    }
                }
            }
        }

        return result;
    }
}